# 4.1.
如图的二次四边形单元，试计算$\frac{\partial{N_1}}{\partial{x}}$和$\frac{\partial{N_2}}{\partial{y}}$
在自然坐标为$(1/2,1/2)$的点的数值 。

In [13]:
import sympy as sp
from sympy import Matrix, Rational
from sympy.abc import xi, eta, x, y
from finite_element.interpolation import Serendipity, SurfaceExclude, Interpolation, Isoparametric
sp.init_printing()


## 计算插值函数

In [2]:
surface_exclude = SurfaceExclude(axises=[xi, eta])
surfaces = [xi+1, xi-1, eta+1, eta-1]
local_coordinates = [[1, 1], [-1, 1], [-1, -1], [1, -1]]
[surface_exclude.add_surface(surface) for surface in surfaces]
interpolation_functions = [
    surface_exclude.get_interpolation_function(local_coordinate)
    for local_coordinate in local_coordinates]
Matrix(interpolation_functions)


⎡ (η + 1)⋅(ξ + 1) ⎤
⎢ ─────────────── ⎥
⎢        4        ⎥
⎢                 ⎥
⎢-(η + 1)⋅(ξ - 1) ⎥
⎢─────────────────⎥
⎢        4        ⎥
⎢                 ⎥
⎢ (η - 1)⋅(ξ - 1) ⎥
⎢ ─────────────── ⎥
⎢        4        ⎥
⎢                 ⎥
⎢-(η - 1)⋅(ξ + 1) ⎥
⎢─────────────────⎥
⎣        4        ⎦

## 计算等参元的雅克比矩阵

整体坐标下的各点位坐标

In [5]:
global_coordinates = [
    [40, 50],
    [5, 40],
    [10, 10],
    [30, 20],
]
Matrix(global_coordinates)

⎡40  50⎤
⎢      ⎥
⎢5   40⎥
⎢      ⎥
⎢10  10⎥
⎢      ⎥
⎣30  20⎦

In [10]:
isoparametric = Isoparametric(
    local_axises=[xi, eta],
    global_axises=[x, y],
    interpolation_functions=interpolation_functions,
    coordinates=global_coordinates)
isoparametric.jacobian

⎡15⋅η   55    ⎤
⎢──── + ──  5 ⎥
⎢ 4     4     ⎥
⎢             ⎥
⎢15⋅ξ   5     ⎥
⎢──── + ─   15⎥
⎣ 4     4     ⎦

## 计算等参元对整体坐标的偏导数

In [11]:
isoparametric.displacement_interpolation_function_diff_global

⎡   3⋅η - ξ + 2        -(3⋅η - ξ + 4)        3⋅η - ξ - 2        -(3⋅η - ξ - 4)
⎢──────────────────   ──────────────────  ──────────────────   ───────────────
⎢5⋅(9⋅η - 3⋅ξ + 32)   5⋅(9⋅η - 3⋅ξ + 32)  5⋅(9⋅η - 3⋅ξ + 32)   5⋅(9⋅η - 3⋅ξ + 
⎢                                                                             
⎢    η + 4⋅ξ + 5         η - 2⋅ξ + 3       -(2⋅η - 7⋅ξ + 5)      -(η + 7⋅ξ + 6
⎢───────────────────  ──────────────────  ───────────────────  ───────────────
⎣10⋅(9⋅η - 3⋅ξ + 32)  5⋅(9⋅η - 3⋅ξ + 32)  10⋅(9⋅η - 3⋅ξ + 32)  10⋅(9⋅η - 3⋅ξ +

    ⎤
─── ⎥
32) ⎥
    ⎥
)   ⎥
────⎥
 32)⎦

代入求值

In [14]:
isoparametric.displacement_interpolation_function_diff_global.subs({xi: Rational(1, 2), eta: Rational(1, 2)})

⎡3/175  -1/35  -1/175  3/175⎤
⎢                           ⎥
⎣3/140  1/70   -1/140  -1/35⎦

其中，整体坐标下，$\frac{\partial{N_1}}{\partial{x}}=\frac{3}{175}$和$\frac{\partial{N_2}}{\partial{y}}=\frac{1}{70}$

库代码见[这里](https://github.com/panhaoyu/finite_element)